DSharpPlus.Core/Rest/Ratelimiting/PollyRatelimitingPolicy.cs

```diff
@@ -0,0 +1,172 @@
+            }
+
+            // ensure a valid cache was passed. the cache mostly serves as an alternative to v4's "BucketCleanerTask",
+            // which, in its static manner, would hardly work with the very DI-focused architecture of Polly and Polly policies.
+            if (!context.TryGetValue("cache", out object cacheRaw) || cacheRaw is not IMemoryCache cache)
+            {
+                throw new InvalidOperationException("No cache passed.");
+            }
+
+            // avoid having to create multiple of these. execution time should be reasonably fast.
+            DateTimeOffset instant = DateTimeOffset.UtcNow;
+
+            // some endpoints are exempt from the global limit, account for them.
+            bool subjectToGlobalLimit = true;
+
+            // this is our unique identifier for each active bucket.
+            string bucketHash;
+
+            // make sure to reflect context-passed exemption
+            if (context.TryGetValue("subject-to-global-limit", out object globallyLimitedRaw) && globallyLimitedRaw is bool globallyLimited)
+            {
+                subjectToGlobalLimit = globallyLimited;
+            }
+
+            // the following logic applies only to globally ratelimited requests.
+            if (subjectToGlobalLimit)
+            {
+                // if the bucket is already expired, reset it. technically we could save a bit of time by goto-ing out of the
+                // if clause entirely if this check succeeds, since we are then guaranteed the request will be allowed.
+                if (globalBucket.ResetTime < instant)
+                {
+                    globalBucket.ResetBucket(instant + oneSecond);
+                }
+
+                if (!globalBucket.AllowNextRequest())
+                {
+                    HttpResponseMessage response = new(HttpStatusCode.TooManyRequests);
+
+                    response.Headers.RetryAfter = new(globalBucket.ResetTime - instant);
+
+                    // allow users to figure out whether the 429 showing up in their debug windows was thrown by Discord or
+                    // an internal DSharpPlus response. also specify what bucket denied the request.
+                    response.Headers.Add("X-DSharpPlus-Preemptive-Ratelimit", "global");
+
+                    // pretend this was a discord api call and return the response.
+                    return response;
+                }
+            }
+
+            // if we don't already have this bucket cached, set the hash to our endpoint for now. 
+            // this will later be used to identify a newly called bucket.
+            bucketHash = endpointBuckets.TryGetValue(endpoint, out string? nullableBucketHash)
+                ? nullableBucketHash
+                : endpoint;
+
+            // get the cached ratelimit bucket. if it doesn't exist, we are guaranteed our request will be allowed
+            // as it will be the first request to this bucket, which never 429s.
```

Comment:
We are not; buckets are stored in memory so a restart could lead to unexpectedly being 429'd, but that's for the user to deal with.

Reply:
Good point, but I don't think there's much we can do about that anyway.